ramips: add support to Linksys EA7300 v1 & v2 boards

rodsan0 opened the pull request:
### Overview

This PR adds support for Linksys EA7300 boards. Given their similarities to the already-supported EA7500 (they are basically the same router) this was surprisingly easy. Support for both versions is included under the same device given that they only differ in their 2.4 GHz radio. Both radios are supported by the same driver so this should not be a problem.

### Specifications

* SoC:  MediaTek MT7621A                (880 MHz 2c/4t)
* RAM:
  * v1: Nanya NT5CC128M16IP-DIT     (256M DDR3-1600)
  * v2: Winbond W632GG6MB-12        (256M DDR3-1600)
* Flash:
  * v1: Macronix MX30LF1G18AC-TI    (128M NAND)
  * v2: Winbond W29N01HVSINA        (128M NAND)
* Eth:  MediaTek MT7621A                (10/100/1000 Mbps x5)
* Radio:
  * v1: MT7615N                     (2.4 GHz & 5 GHz)
  * v2:
    * MT7603E                     (2.4 GHz)
    * MT7615N                     (5 GHz)
  * 4 antennae: 1 internal and 3 non-deatachable
* USB:  3.0 (x1)
* LEDs:
  * White   (x1 logo)
  * Green   (x6 eth + wps)
  * Orange  (x5, hardware-bound)
* Buttons:
  * Reset   (x1)
  * WPS     (x1)

Everything works! Been running it for a couple weeks now and haven't had
any problems.

### Installation

Flash factory image through GUI.

This might fail due to the A/B nature of this device. When flashing, OEM
firmware writes over the non-booted partition. If booted from 'A',
flashing over 'B' won't work. To get around this, you should flash the
OEM image over itself. This will then boot the router from 'B' and
allow you to flash OpenWRT without problems.

### Reverting to factory firmware

Hard-reset the router three times to force it to boot from 'B.' This is
where the stock firmware resides. To remove any traces of OpenWRT from
your router simply flash the OEM image at this point.

adrianschmutzler commented:
> Been running it for a couple weeks now

Does "it" mean you only have one (vX) of them?

adrianschmutzler commented:
As it appears the DTS file is identical to ea7500-v2, please create a shared DTSI for them (either mt7621_linksys_ea7xxx.dtsi or based on some board codename if known). The DTS files for both ea7500-v2 and your device(s) would only contain compatible/model, the aliases, and the LED block then.

adrianschmutzler commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -544,6 +544,23 @@ define Device/linksys_ea7500-v2
+  KERNEL_SIZE := 4096k
+  IMAGE_SIZE := 36864k
+  DEVICE_VENDOR := Linksys
+  DEVICE_MODEL := EA7300
```
How probable is it that there will be a v3 (with different setup)?

Having linksys_ea7300 and linksys_ea7300-v3 would not be so helpful, then I'd consider to split and explicitly define ea7300-v1 and -v2 already.

rodsan0 replied:
You're right, I separated both versions into their own devices.

adrianschmutzler commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -544,6 +544,23 @@ define Device/linksys_ea7500-v2
 endef
 TARGET_DEVICES += linksys_ea7500-v2
 
+define Device/linksys_ea7300
```
This whole block is mostly a repetition of the device before. Please create a shared definition as e.g. with Device/netgear_sercomm_nand and then inherit the devices from it.
Despite, ea7300 should come _before_ ea7500.

adrianschmutzler commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -544,6 +544,23 @@ define Device/linksys_ea7500-v2
+  IMAGES := sysupgrade.bin factory.bin
+  IMAGE/sysupgrade.bin := sysupgrade-tar | append-metadata | check-size
+  IMAGE/factory.bin := append-kernel | pad-to $$$$(KERNEL_SIZE) | \
+	append-ubi | check-size | linksys-image type=EA7300
```
One could consider introducing a variable for the type here, as it's done with SERCOMM_HWNAME in Device/netgear_sercomm_nand. Then, the rest of the statement would be the same for the devices, and could be moved to the common definition as well.

adrianschmutzler commented on target/linux/ramips/mt7621/base-files/etc/board.d/01_leds:
```diff
@@ -35,7 +35,8 @@ gnubee,gb-pc2)
 	;;
-linksys,ea7500-v2)
+linksys,ea7500-v2|\
+linksys,ea7300)
```
Sorting.

adrianschmutzler commented on target/linux/ramips/mt7621/base-files/etc/board.d/02_network:
```diff
@@ -100,7 +100,8 @@ ramips_setup_macs()
 		;;
-	linksys,ea7500-v2)
+	linksys,ea7500-v2|\
+	linksys,ea7300)
```
Sorting.

adrianschmutzler commented on target/linux/ramips/mt7621/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac:
```diff
@@ -10,7 +10,8 @@ PHYNBR=${DEVPATH##*/phy}
 case "$board" in
-	linksys,ea7500-v2)
+	linksys,ea7500-v2|\
+	linksys,ea7300)
```
Sorting.

adrianschmutzler commented on target/linux/ramips/mt7621/base-files/etc/init.d/bootcount:
```diff
@@ -8,7 +8,8 @@ boot() {
 		;;
-	linksys,ea7500-v2)
+	linksys,ea7500-v2|\
+	linksys,ea7300)
```
Sorting.

adrianschmutzler commented on target/linux/ramips/mt7621/base-files/lib/upgrade/platform.sh:
```diff
@@ -46,6 +46,7 @@ platform_do_upgrade() {
 	asus,rt-ac85p|\
 	hiwifi,hc5962|\
 	linksys,ea7500-v2|\
+	linksys,ea7300|\
```
Sorting.

adrianschmutzler commented on package/boot/uboot-envtools/files/ramips:
```diff
@@ -34,6 +34,7 @@ zbtlink,zbt-wg2626)
 	ubootenv_add_uci_config "/dev/mtd1" "0x0" "0x1000" "0x10000"
 	;;
 linksys,ea7500-v2|\
+linksys,ea7300|\
```
Sorting.

adrianschmutzler commented:
Seem to be in good shape, now just let's wrap it up properly.

rodsan0 commented:
> > Been running it for a couple weeks now
> 
> Does "it" mean you only have one (vX) of them?

I only own v1. I had a friend with v2 install it but I sadly couldn't personally test it. Should I make a note of this somewhere?

rodsan0 committed:
ramips: add support to Linksys EA7300 v1 & v2 boards

Specifications:

* SoC:  MediaTek MT7621A                (880 MHz 2c/4t)
* RAM:
        v1: Nanya NT5CC128M16IP-DIT     (256M DDR3-1600)
        v2: Winbond W632GG6MB-12        (256M DDR3-1600)
* Flash:
        v1: Macronix MX30LF1G18AC-TI    (128M NAND)
        v2: Winbond W29N01HVSINA        (128M NAND)
* Eth:  MediaTek MT7621A                (10/100/1000 Mbps x5)
* Radio:
        v1: MT7615N                     (2.4 GHz & 5 GHz)
        v2: MT7603E                     (2.4 GHz)
            MT7615N                     (5 GHz)
        4 antennae: 1 internal and 3 non-deatachable
* USB:  3.0 (x1)
* LEDs:
    White   (x1 logo)
    Green   (x6 eth + wps)
    Orange  (x5, hardware-bound)
* Buttons:
    Reset   (x1)
    WPS     (x1)

Everything works! Been running it for a couple weeks now and haven't had
any problems. Please let me know if you run into any.

Installation:

Flash factory image through GUI.

This might fail due to the A/B nature of this device. When flashing, OEM
firmware writes over the non-booted partition. If booted from 'A',
flashing over 'B' won't work. To get around this, you should flash the
OEM image over itself. This will then boot the router from 'B' and
allow you to flash OpenWRT without problems.

Reverting to factory firmware:

Hard-reset the router three times to force it to boot from 'B.' This is
where the stock firmware resides. To remove any traces of OpenWRT from
your router simply flash the OEM image at this point.

Signed-off-by: Santiago Rodriguez-Papa <contact@rodsan.dev>

rodsan0 commented:
I separated both versions into their own devices. 

After double-checking the OEM images it seems that the hardware type diverges for v2. This didn't seem to affect flashing but is now fixed nonetheless. We are now also packaging kmod-mt7603 for v2 but not v1, since the later doesn't need it anyways.

Implementing a common definition for EA7xxx devices will also make it easier to add support for the EA7400 and EA7450 at a later time.

After these changes, EA7300 v1, v2, and EA7500 make without error, and v1 runs fine.

adrianschmutzler commented:
I've decided that it might be better to add devices separately after looking at your combined commit message.
I've cut out v1 from your patch and pushed it with some (minor) adjustments to master already.

So, you could now provide a separate patch for v2 and push it here (easiest by just rebasing your stuff onto master and adjust the commit message to v2 only ...).

adrianschmutzler commented:
To keep to old state conserved here, I'm closing this PR now.
v1 has been merged, please submit a separate patch for v2 in a separate PR (should be easy to make from what's here already).
